**1. check quickly that tensorflow is there in Saturn CLoud**

In [1]:
import tensorflow as tf

2022-11-23 14:10:49.469170: I tensorflow/core/util/util.cc:169] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.


In [2]:
tf.__version__

'2.9.1'

## Homework 08

### Dataset

In this homework, we'll build a model for predicting if we have an image of a dino or a dragon. For this, we will use the "Dino or Dragon?" dataset that can be downloaded from Kaggle.

You can get a wget-able version here:

`wget https://github.com/alexeygrigorev/dino-or-dragon/releases/download/data/dino-dragon.zip`

`unzip dino-dragon.zip`

In the lectures we saw how to use a pre-trained neural network. In the homework, we'll train a much smaller model from scratch.

**Note:** You will need an environment with a GPU for this homework. We recommend to use [Saturn Cloud](https://bit.ly/saturn-mlzoomcamp). You can also use a computer without a GPU (e.g. your laptop), but it will be slower.

### Data Preparation

The dataset contains around 1900 images of dinos and around 1900 images of dragons.

The dataset contains separate folders for training and test sets.

## 2. Getting the images for homework

In [3]:
!wget https://github.com/alexeygrigorev/dino-or-dragon/releases/download/data/dino-dragon.zip

--2022-11-22 15:54:47--  https://github.com/alexeygrigorev/dino-or-dragon/releases/download/data/dino-dragon.zip
Resolving github.com (github.com)... 140.82.114.4
Connecting to github.com (github.com)|140.82.114.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/540945974/eb853cee-1dfd-473e-a770-9527ef502079?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20221122%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20221122T155447Z&X-Amz-Expires=300&X-Amz-Signature=e712756b8a6cc0db6258ec76764029b78a835d6f45e5ef3aa4e3e4f49194de9a&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=540945974&response-content-disposition=attachment%3B%20filename%3Ddino-dragon.zip&response-content-type=application%2Foctet-stream [following]
--2022-11-22 15:54:47--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/540945974/eb853cee-1dfd-473e-a770-9527ef5020

In [4]:
!unzip dino-dragon.zip

Archive:  dino-dragon.zip
  inflating: test/dino/001db8b8-33ff-44b3-8f50-325f4ab4f713.jpg  
  inflating: test/dino/009d2929-87f4-43d8-8b7d-501d8a006823.jpg  
  inflating: test/dino/00f2e69c-9586-4b06-a306-8fb36b88b651.jpg  
  inflating: test/dino/0289c104-155f-4e8e-99bf-8b6129cef3e7.jpg  
  inflating: test/dino/03b62646-cdd1-4988-b3ca-d13afd91e34b.jpg  
  inflating: test/dino/04bd253a-f460-4861-99aa-93e48cf6229f.jpg  
  inflating: test/dino/073500e1-e6f7-4274-bf5a-7c617f0c8d79.jpg  
  inflating: test/dino/073b18b9-3e1d-4659-b19c-e9daa61bc391.jpg  
  inflating: test/dino/09153010-3dc0-482e-9217-533341735aba.jpg  
  inflating: test/dino/09cdfe5c-a6c2-4d8e-bd09-15797b54a656.jpg  
  inflating: test/dino/0c1d462a-d923-4232-befd-b29be94338b2.jpg  
  inflating: test/dino/0c494627-2ed2-49c7-9ff7-e70308ad5fcb.jpg  
  inflating: test/dino/0d50f95a-8a4c-4712-8187-2f4008fd6241.jpg  
  inflating: test/dino/0e985370-79bc-456c-b0aa-c488d63dfa4f.jpg  
  inflating: test/dino/131e8d9c-01b6-49e8-9d4f-1cc

# Model

For this homework we will use Convolutional Neural Network (CNN). Like in the lectures, we'll use Keras.

You need to develop the model with following structure:


* The shape for input should be `(150, 150, 3)`
* Next, create a convolutional layer ([`Conv2D`](https://keras.io/api/layers/convolution_layers/convolution2d/)):
    * Use 32 filters
    * Kernel size should be `(3, 3)` (that's the size of the filter)
    * Use `'relu'` as activation 
* Reduce the size of the feature map with max pooling ([`MaxPooling2D`](https://keras.io/api/layers/pooling_layers/max_pooling2d/))
    * Set the pooling size to `(2, 2)`
* Turn the multi-dimensional result into vectors using a [`Flatten`](https://keras.io/api/layers/reshaping_layers/flatten/) layer
* Next, add a `Dense` layer with 64 neurons and `'relu'` activation
* Finally, create the `Dense` layer with 1 neuron - this will be the output
    * The output layer should have an activation - use the appropriate activation for the binary classification case

As optimizer use [`SGD`](https://keras.io/api/optimizers/sgd/) with the following parameters:

* `SGD(lr=0.002, momentum=0.8)`

For clarification about kernel size and max pooling, check [Office Hours](https://www.youtube.com/watch?v=1WRgdBTUaAc).


In [3]:
import numpy as np
import matplotlib.pyplot as plt

%matplotlib inline

In [4]:
import tensorflow as tf
from tensorflow import keras

from tensorflow.keras import models
from tensorflow.keras import layers
from tensorflow.keras import optimizers

## Developing the model

In [5]:
#https://www.tensorflow.org/guide/keras/save_and_serialize

model = models.Sequential()

model.add(layers.Conv2D(32, (3, 3), activation='relu',
                        input_shape=(150, 150, 3)))
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Flatten())
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))

model.compile(loss='binary_crossentropy',\
             optimizer=optimizers.SGD(learning_rate=0.002, momentum=0.8),\
             metrics=['acc'])

2022-11-23 14:12:16.308611: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:975] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-11-23 14:12:16.385885: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:975] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-11-23 14:12:16.386551: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:975] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2022-11-23 14:12:16.388940: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F AVX512_VNNI FMA
To enable them in other operations, rebuild TensorFlow with the approp

In [6]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 conv2d (Conv2D)             (None, 148, 148, 32)      896       
                                                                 
 max_pooling2d (MaxPooling2D  (None, 74, 74, 32)       0         
 )                                                               
                                                                 
 flatten (Flatten)           (None, 175232)            0         
                                                                 
 dense (Dense)               (None, 64)                11214912  
                                                                 
 dense_1 (Dense)             (None, 1)                 65        
                                                                 
Total params: 11,215,873
Trainable params: 11,215,873
Non-trainable params: 0
____________________________________________

### Question 1

Since we have a binary classification problem, what is the best loss function for us?

- `binary crossentropy`
- `focal loss`
- `mean squared error`
- `categorical crossentropy`

Note: since we specify an activation for the output layer, we don't need to set `from_logits=True`

**Answer to question 1**

`binray crossentropy`

### Question 2

What's the total number of parameters of the model? You can use the `summary` method for that. 

- 9215873
- 11215873
- 14215873
- 19215873

**Answer to question 2**

`11215873`

### Generators and Training

For the next two questions, use the following data generator for both train and test sets:

```python
ImageDataGenerator(rescale=1./255)
```

* We don't need to do any additional pre-processing for the images.
* When reading the data from train/test directories, check the `class_mode` parameter. Which value should it be for a binary classification problem?
* Use `batch_size=20`
* Use `shuffle=True` for both training and test sets. 

For training use `.fit()` with the following params:

```python
model.fit(
    train_generator,
    epochs=10,
    validation_data=test_generator
)
```

In [10]:
from tensorflow.keras.preprocessing.image import ImageDataGenerator

In [11]:
train_gen = ImageDataGenerator(rescale=1./255)
test_gen = ImageDataGenerator(rescale=1./255)

In [12]:
ls

dino-dragon.zip                       README.md                    test/
hw08-mlzoomcamp-dino-or-dragon.ipynb  single-gpu-tensorflow.ipynb  train/


In [13]:
train_dir = './train'
test_dir = './test'

In [14]:
train_ds = train_gen.flow_from_directory(train_dir,
                                        target_size=(150, 150),
                                        batch_size=20,
                                        class_mode='binary',
                                        shuffle = True)

Found 1594 images belonging to 2 classes.


In [15]:
test_ds = test_gen.flow_from_directory(test_dir,
                                       target_size=(150, 150),
                                       batch_size=20,
                                       class_mode='binary',
                                       shuffle = True)

Found 394 images belonging to 2 classes.


In [16]:
#training
history = model.fit(
    train_ds,
    epochs=10,
    validation_data=test_ds,
)

Epoch 1/10


2022-11-23 14:17:45.230506: I tensorflow/stream_executor/cuda/cuda_dnn.cc:384] Loaded cuDNN version 8100
2022-11-23 14:17:47.230885: I tensorflow/core/platform/default/subprocess.cc:304] Start cannot spawn child process: No such file or directory
2022-11-23 14:17:47.231777: I tensorflow/core/platform/default/subprocess.cc:304] Start cannot spawn child process: No such file or directory
2022-11-23 14:17:47.231816: W tensorflow/stream_executor/gpu/asm_compiler.cc:80] Couldn't get ptxas version string: INTERNAL: Couldn't invoke ptxas --version
2022-11-23 14:17:47.232804: I tensorflow/core/platform/default/subprocess.cc:304] Start cannot spawn child process: No such file or directory
2022-11-23 14:17:47.232929: W tensorflow/stream_executor/gpu/redzone_allocator.cc:314] INTERNAL: Failed to launch ptxas
Relying on driver to perform ptx compilation. 
Modify $PATH to customize ptxas location.
This message will be only logged once.


80/80 [==============================] - 16s 123ms/step - loss: 0.6861 - acc: 0.5621 - val_loss: 0.6590 - val_acc: 0.7437
Epoch 2/10
80/80 [==============================] - 9s 107ms/step - loss: 0.6236 - acc: 0.6863 - val_loss: 0.5655 - val_acc: 0.7843
Epoch 3/10
80/80 [==============================] - 9s 106ms/step - loss: 0.5130 - acc: 0.7823 - val_loss: 0.5343 - val_acc: 0.7132
Epoch 4/10
80/80 [==============================] - 8s 105ms/step - loss: 0.4333 - acc: 0.8149 - val_loss: 0.4783 - val_acc: 0.7665
Epoch 5/10
80/80 [==============================] - 9s 106ms/step - loss: 0.3717 - acc: 0.8463 - val_loss: 0.4640 - val_acc: 0.7817
Epoch 6/10
80/80 [==============================] - 8s 106ms/step - loss: 0.3347 - acc: 0.8689 - val_loss: 0.3494 - val_acc: 0.8426
Epoch 7/10
80/80 [==============================] - 8s 106ms/step - loss: 0.3067 - acc: 0.8783 - val_loss: 0.3356 - val_acc: 0.8503
Epoch 8/10
80/80 [==============================] - 9s 107ms/step - loss: 0.2689 - acc

In [17]:
acc = history.history['acc']
acc_median = np.median(acc)
acc_median

0.8575909733772278

In [18]:
loss = history.history['loss']
loss_std = np.std(loss)
loss_std

0.15222313076480457

### Question 3

What is the median of training accuracy for all the epochs for this model?

- 0.40
- 0.60
- 0.90
- 0.20

**Answer to question 3**

`0.90` (closest answer to 0.86)

### Question 4

What is the standard deviation of training loss for all the epochs for this model?

- 0.11
- 0.66
- 0.99
- 0.33

**Answer to question 4**
`0.11` (closest answer to my answer 0.15)

## Data Augmentation

For the next two questions, we'll generate more data using data augmentations. 

Add the following augmentations to your training data generator:

* `rotation_range=40,`
* `width_shift_range=0.2,`
* `height_shift_range=0.2,`
* `shear_range=0.2,`
* `zoom_range=0.2,`
* `horizontal_flip=True,`
* `fill_mode='nearest'`

In [21]:
datagen = ImageDataGenerator(
    rotation_range=40,
    width_shift_range=0.2,
    height_shift_range=0.2,
    shear_range=0.2,
    zoom_range=0.2,
    horizontal_flip=True,
    fill_mode='nearest')

In [22]:
from tensorflow.keras.preprocessing import image

In [23]:
train_datagen = ImageDataGenerator(
    rescale=1./255,
    rotation_range=40,
    width_shift_range=0.2,
    height_shift_range=0.2,
    shear_range=0.2,
    zoom_range=0.2,
    horizontal_flip=True,)

In [24]:
test_datagen = ImageDataGenerator(rescale=1./255)

In [25]:
train_generator = train_datagen.flow_from_directory(train_dir,
                                                    target_size=(150, 150), 
                                                    batch_size=32, 
                                                    class_mode='binary')

Found 1594 images belonging to 2 classes.


In [26]:
test_generator = test_datagen.flow_from_directory(
    test_dir,
    target_size=(150, 150),
    batch_size=32,
    class_mode='binary')

Found 394 images belonging to 2 classes.


In [28]:
# training
history = model.fit(
    train_generator,
    epochs=10,
    validation_data=test_generator
)

Epoch 1/10
50/50 [==============================] - 15s 306ms/step - loss: 0.4309 - acc: 0.8018 - val_loss: 0.3900 - val_acc: 0.8223
Epoch 2/10
50/50 [==============================] - 15s 301ms/step - loss: 0.4298 - acc: 0.7992 - val_loss: 0.4365 - val_acc: 0.7944
Epoch 3/10
50/50 [==============================] - 15s 301ms/step - loss: 0.4027 - acc: 0.8237 - val_loss: 0.3718 - val_acc: 0.8299
Epoch 4/10
50/50 [==============================] - 15s 301ms/step - loss: 0.4083 - acc: 0.8174 - val_loss: 0.3257 - val_acc: 0.8579
Epoch 5/10
50/50 [==============================] - 15s 303ms/step - loss: 0.4018 - acc: 0.8156 - val_loss: 0.3262 - val_acc: 0.8629
Epoch 6/10
50/50 [==============================] - 15s 298ms/step - loss: 0.3969 - acc: 0.8231 - val_loss: 0.3050 - val_acc: 0.8604
Epoch 7/10
50/50 [==============================] - 15s 303ms/step - loss: 0.3897 - acc: 0.8206 - val_loss: 0.4487 - val_acc: 0.7995
Epoch 8/10
50/50 [==============================] - 15s 301ms/step - 

### Question 5 

Let's train our model for 10 more epochs using the same code as previously.
Make sure you don't re-create the model - we want to continue training the model
we already started training.

What is the mean of test loss for all the epochs for the model trained with augmentations?

- 0.15
- 0.77
- 0.37
- 0.97

In [29]:
val_loss_aug = history.history['val_loss']
loss_mean_aug = np.mean(val_loss_aug)
loss_mean_aug

0.3685357987880707

### Question 6

What's the average of test accuracy for the last 5 epochs (from 6 to 10)
for the model trained with augmentations?

- 0.84
- 0.54
- 0.44
- 0.24

In [30]:
val_acc_aug = history.history['val_acc']
acc_mean_aug = np.mean(val_acc_aug[5:10])
acc_mean_aug

0.8401015162467956

## Submit the results

- Submit your results here: https://forms.gle/XdH5ztBddvTvxzpT6
- You can submit your solution multiple times. In this case, only the last submission will be used
- If your answer doesn't match options exactly, select the closest one


## Deadline

The deadline for submitting is 21 November 2022, 23:00 CEST.

After that, the form will be closed.